In [1]:
import pandas as pd 
import numpy as np


In [2]:
df = pd.read_csv("telco.csv") #reading the files 

In [3]:
df.head(10) 

,Customer ID,Gender,Age,Under 30,Senior Citizen,Married,Dependents,Number of Dependents,Country,State,...,Total Extra Data Charges,Total Long Distance Charges,Total Revenue,Satisfaction Score,Customer Status,Churn Label,Churn Score,CLTV,Churn Category,Churn Reason
0,8779-QRDMV,Male,78,No,Yes,No,No,0,United States,California,...,20,0.00,59.65,3,Churned,Yes,91,5433,Competitor,Competitor offered more data
1,7495-OOKFY,Female,74,No,Yes,Yes,Yes,1,United States,California,...,0,390.80,1024.10,3,Churned,Yes,69,5302,Competitor,Competitor made better offer
2,1658-BYGOY,Male,71,No,Yes,No,Yes,3,United States,California,...,0,203.94,1910.88,2,Churned,Yes,81,3179,Competitor,Competitor made better offer
3,4598-XLKNJ,Female,78,No,Yes,Yes,Yes,1,United States,California,...,0,494.00,2995.07,2,Churned,Yes,88,5337,Dissatisfaction,Limited range of services
4,4846-WHAFZ,Female,80,No,Yes,Yes,Yes,1,United States,California,...,0,234.21,3102.36,2,Churned,Yes,67,2793,Price,Extra data charges
5,4412-YLTKF,Female,72,No,Yes,No,Yes,1,United States,California,...,10,89.91,2235.41,1,Churned,Yes,95,4638,Competitor,Competitor had better devices
6,0390-DCFDQ,Female,76,No,Yes,Yes,Yes,2,United States,California,...,0,15.28,85.73,2,Churned,Yes,76,3964,Other,Don't know
7,3445-HXXGF,Male,66,No,Yes,Yes,No,0,United States,California,...,0,0.00,2610.25,1,Churned,Yes,91,5444,Dissatisfaction,Service dissatisfaction
8,2656-FMOKZ,Female,70,No,Yes,No,Yes,2,United States,California,...,0,661.05,1806.75,2,Churned,Yes,91,5717,Dissatisfaction,Limited range of services
9,2070-FNEXE,Female,77,No,Yes,No,Yes,2,United States,California,...,0,188.65,681.20,2,Churned,Yes,81,4419,Price,Lack of affordable download/upload speed


In [5]:
df.shape

(7043, 50)

In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 50 columns):
 #   Column                             Non-Null Count  Dtype  
---  ------                             --------------  -----  
 0   Customer ID                        7043 non-null   object 
 1   Gender                             7043 non-null   object 
 2   Age                                7043 non-null   int64  
 3   Under 30                           7043 non-null   object 
 4   Senior Citizen                     7043 non-null   object 
 5   Married                            7043 non-null   object 
 6   Dependents                         7043 non-null   object 
 7   Number of Dependents               7043 non-null   int64  
 8   Country                            7043 non-null   object 
 9   State                              7043 non-null   object 
 10  City                               7043 non-null   object 
 11  Zip Code                           7043 non-null   int64

In [13]:
df.describe()

,Age,Number of Dependents,Zip Code,Latitude,Longitude,Population,Number of Referrals,Tenure in Months,Avg Monthly Long Distance Charges,Avg Monthly GB Download,Monthly Charge,Total Charges,Total Refunds,Total Extra Data Charges,Total Long Distance Charges,Total Revenue,Satisfaction Score,Churn Score,CLTV
count,7043.000000,7043.000000,7043.000000,7043.000000,7043.000000,7043.000000,7043.000000,7043.000000,7043.000000,7043.000000,7043.000000,7043.000000,7043.000000,7043.000000,7043.000000,7043.000000,7043.000000,7043.000000,7043.000000
mean,46.509726,0.468692,93486.070567,36.197455,-119.756684,22139.603294,1.951867,32.386767,22.958954,20.515405,64.761692,2280.381264,1.962182,6.860713,749.099262,3034.379056,3.244924,58.505040,4400.295755
std,16.750352,0.962802,1856.767505,2.468929,2.154425,21152.392837,3.001199,24.542061,15.448113,20.418940,30.090047,2266.220462,7.902614,25.104978,846.660055,2865.204542,1.201657,21.170031,1183.057152
min,19.000000,0.000000,90001.000000,32.555828,-124.301372,11.000000,0.000000,1.000000,0.000000,0.000000,18.250000,18.800000,0.000000,0.000000,0.000000,21.360000,1.000000,5.000000,2003.000000
25%,32.000000,0.000000,92101.000000,33.990646,-121.788090,2344.000000,0.000000,9.000000,9.210000,3.000000,35.500000,400.150000,0.000000,0.000000,70.545000,605.610000,3.000000,40.000000,3469.000000
50%,46.000000,0.000000,93518.000000,36.205465,-119.595293,17554.000000,0.000000,29.000000,22.890000,17.000000,70.350000,1394.550000,0.000000,0.000000,401.440000,2108.640000,3.000000,61.000000,4527.000000
75%,60.000000,0.000000,95329.000000,38.161321,-117.969795,36125.000000,3.000000,55.000000,36.395000,27.000000,89.850000,3786.600000,0.000000,0.000000,1191.100000,4801.145000,4.000000,75.500000,5380.500000
max,80.000000,9.000000,96150.000000,41.962127,-114.192901,105285.000000,11.000000,72.000000,49.990000,85.000000,118.750000,8684.800000,49.790000,150.000000,3564.720000,11979.340000,5.000000,96.000000,6500.000000


In [15]:
df.describe(include="object")

,Customer ID,Gender,Under 30,Senior Citizen,Married,Dependents,Country,State,City,Quarter,...,Streaming Movies,Streaming Music,Unlimited Data,Contract,Paperless Billing,Payment Method,Customer Status,Churn Label,Churn Category,Churn Reason
count,7043,7043,7043,7043,7043,7043,7043,7043,7043,7043,...,7043,7043,7043,7043,7043,7043,7043,7043,1869,1869
unique,7043,2,2,2,2,2,1,1,1106,1,...,2,2,2,3,2,3,3,2,5,20
top,8779-QRDMV,Male,No,No,No,No,United States,California,Los Angeles,Q3,...,No,No,Yes,Month-to-Month,Yes,Bank Withdrawal,Stayed,No,Competitor,Competitor had better devices
freq,1,3555,5642,5901,3641,5416,7043,7043,293,7043,...,4311,4555,4745,3610,4171,3909,4720,5174,841,313


In [23]:
df.isnull().sum().sort_values(ascending=False)

Churn Reason                         5174
Churn Category                       5174
Offer                                3877
Internet Type                        1526
Payment Method                          0
Device Protection Plan                  0
Premium Tech Support                    0
Streaming TV                            0
Streaming Movies                        0
Streaming Music                         0
Unlimited Data                          0
Contract                                0
Paperless Billing                       0
Monthly Charge                          0
Online Security                         0
Total Charges                           0
Total Refunds                           0
Total Extra Data Charges                0
Total Long Distance Charges             0
Total Revenue                           0
Satisfaction Score                      0
Customer Status                         0
Churn Label                             0
Churn Score                       

The dataset contains Four columns that has missing values:Offer, Internet Type, Churn Category and Churn Reason.
Initial investigation suggests that some missing values are not data quality issues but rather represent business conditions. For example, churn-related fields are expected to be blank for customers who have not churned.

In [19]:
df["Customer Status"].value_counts()

Customer Status
Stayed     4720
Churned    1869
Joined      454
Name: count, dtype: int64

In [21]:
df[df["Churn Category"].isnull()]["Customer Status"].value_counts()

Customer Status
Stayed    4720
Joined     454
Name: count, dtype: int64

In [18]:
df[df["Churn Reason"].isnull()]["Customer Status"].value_counts()

Customer Status
Stayed    4720
Joined     454
Name: count, dtype: int64

#For Churn Reason and Churn Category         
Missing values in Churn Category and Churn Reason are expected because these variables are only populated for customers who have churned.

Therefore, the missing values represent a valid business condition rather than a data quality issue and were retained.

In [22]:
#For Internet Type
df[df["Internet Type"].isnull()]["Internet Service"].value_counts()

Internet Service
No    1526
Name: count, dtype: int64

In [24]:
df["Internet Type"].value_counts()

Internet Type
Fiber Optic    3035
DSL            1652
Cable           830
Name: count, dtype: int64

Every missing value in the Internet Type may belong to customers who dont have internet service.

In [25]:
df["Internet Type"]=(df["Internet Type"].fillna("No Internet"))
#filling missing values with value=no internet 

In [27]:
df["Internet Type"].value_counts(dropna=False)

Internet Type
Fiber Optic    3035
DSL            1652
No Internet    1526
Cable           830
Name: count, dtype: int64

### Internet Type

The Internet Type variable contained 1,526 missing values.

Investigation revealed that all missing values corresponded to customers whose Internet Service status was "No".

Because these missing values represent a valid business condition rather than a data quality issue, they were replaced with the category "No Internet" to preserve business meaning and improve downstream analysis.

In [28]:
df["Offer"].value_counts(dropna=False)

Offer
NaN        3877
Offer B     824
Offer E     805
Offer D     602
Offer A     520
Offer C     415
Name: count, dtype: int64

In [29]:
df["Offer"]= df["Offer"].fillna("No Offer")

In [30]:
df["Offer"].value_counts(dropna=False)

Offer
No Offer    3877
Offer B      824
Offer E      805
Offer D      602
Offer A      520
Offer C      415
Name: count, dtype: int64

### Offer Variable Investigation

The Offer variable contained 3,877 missing values (55.05% of all records).

Based on the dataset documentation, the Offer variable represents the last marketing offer accepted by a customer, if applicable.

Therefore, missing values indicate that no marketing offer was accepted rather than a data quality issue.

To preserve business meaning, missing values were replaced with the category "No Offer".

In [31]:
df.duplicated().sum()

0

No duplicates found so no changes 

In [33]:
df["Age"].min(), df["Age"].max()

(19, 80)

In [34]:
df["Satisfaction Score"].min(), df["Satisfaction Score"].max()

(1, 5)

In [35]:
df["Churn Score"].min(), df["Churn Score"].max()

(5, 96)

In [36]:
df["Monthly Charge"].describe()

count    7043.000000
mean       64.761692
std        30.090047
min        18.250000
25%        35.500000
50%        70.350000
75%        89.850000
max       118.750000
Name: Monthly Charge, dtype: float64

In [37]:
df["Monthly Charge"].min(), df["Monthly Charge"].max()

(18.25, 118.75)

In [38]:
df[df["Total Revenue"]<0]

,Customer ID,Gender,Age,Under 30,Senior Citizen,Married,Dependents,Number of Dependents,Country,State,...,Total Extra Data Charges,Total Long Distance Charges,Total Revenue,Satisfaction Score,Customer Status,Churn Label,Churn Score,CLTV,Churn Category,Churn Reason


In [39]:
df[df["Total Charges"]<0]

,Customer ID,Gender,Age,Under 30,Senior Citizen,Married,Dependents,Number of Dependents,Country,State,...,Total Extra Data Charges,Total Long Distance Charges,Total Revenue,Satisfaction Score,Customer Status,Churn Label,Churn Score,CLTV,Churn Category,Churn Reason


In [40]:
df["CLTV"].min(), df["CLTV"].max()

(2003, 6500)

### Data Validation

Several key business variables were validated against expected ranges.

Age:  0–100 
Satisfaction Score:  1–5 
Churn Score:  0–100 
Monthly Charge:  Non-negative 
Total Charge: Non-negative
Total Revenue:  Non-negative 
CLTV: 2000-7000

No invalid or out-of-range values were identified during validation. The dataset was deemed suitable for exploratory data analysis and predictive modeling.

In [41]:
df.to_csv("data/processed/cleaned_telco.csv", index=False)

In [43]:
clean_df = pd.read_csv(
    "data/processed/cleaned_telco.csv"
)

clean_df.head()

,Customer ID,Gender,Age,Under 30,Senior Citizen,Married,Dependents,Number of Dependents,Country,State,...,Total Extra Data Charges,Total Long Distance Charges,Total Revenue,Satisfaction Score,Customer Status,Churn Label,Churn Score,CLTV,Churn Category,Churn Reason
0,8779-QRDMV,Male,78,No,Yes,No,No,0,United States,California,...,20,0.00,59.65,3,Churned,Yes,91,5433,Competitor,Competitor offered more data
1,7495-OOKFY,Female,74,No,Yes,Yes,Yes,1,United States,California,...,0,390.80,1024.10,3,Churned,Yes,69,5302,Competitor,Competitor made better offer
2,1658-BYGOY,Male,71,No,Yes,No,Yes,3,United States,California,...,0,203.94,1910.88,2,Churned,Yes,81,3179,Competitor,Competitor made better offer
3,4598-XLKNJ,Female,78,No,Yes,Yes,Yes,1,United States,California,...,0,494.00,2995.07,2,Churned,Yes,88,5337,Dissatisfaction,Limited range of services
4,4846-WHAFZ,Female,80,No,Yes,Yes,Yes,1,United States,California,...,0,234.21,3102.36,2,Churned,Yes,67,2793,Price,Extra data charges



The cleaned dataset was exported as:

data/processed/cleaned_telco.csv

This dataset will be used for exploratory data analysis, statistical testing, machine learning, and Tableau dashboard development.